// tsx-strict@0.2.0/dist/file-watcher.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupFileWatcher = setupFileWatcher;
exports.stopFileWatcher = stopFileWatcher;
const chokidar_1 = __importDefault(require("chokidar"));
const sheu_1 = __importDefault(require("./sheu"));
let fileWatcher = null;
function setupFileWatcher(restartTsx) {
    if (fileWatcher) {
        fileWatcher.close();
        fileWatcher = null;
    }
    fileWatcher = chokidar_1.default.watch([process.cwd()], {
        ignored: [/node_modules/, /(^|[\/\\])\../, /.json/],
        ignoreInitial: true,
        persistent: true,
    });
    fileWatcher.on("all", (event, path) => {
        if (event === "ready" || !/\.(ts|tsx|js|jsx)$/.test(path))
            return;
        path = path.replace(process.cwd(), "");
        path = path.startsWith("/")
            ? path.replace("/", "")
            : path.startsWith("\\")
                ? path.replace("\\", "")
                : path;
        sheu_1.default.info(`Restarting because of file changes: ${path}`, {
            timestamp: true,
        });
        restartTsx?.();
    });
}
function stopFileWatcher() {
    if (fileWatcher) {
        fileWatcher.close();
        fileWatcher = null;
    }
}
