// tsx-strict@0.2.0/dist/index.js
#!/usr/bin/env node
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tsxKiller = void 0;
exports.setTsxKiller = setTsxKiller;
exports.getTsxKiller = getTsxKiller;
exports.runTsxStrict = runTsxStrict;
const node_cleanup_1 = __importStar(require("node-cleanup"));
const cross_spawn_1 = __importDefault(require("cross-spawn"));
const runner_1 = require("./runner");
const stdout_manipulator_1 = require("./stdout-manipulator");
const readline_1 = require("readline");
const killer_1 = require("./killer");
const compiler_provider_1 = require("./compiler-provider");
const file_watcher_1 = require("./file-watcher");
let firstTime = true;
exports.tsxKiller = null;
function setTsxKiller(value) {
    exports.tsxKiller = value;
}
function getTsxKiller() {
    return exports.tsxKiller;
}
async function runTsxStrict(file, options) {
    const { clear = true, typeCheck = true, compiler = "tsc", watch = false, tscArgs = "", tsxArgs = "", maxNodeMem, include, } = options;
    function runTsxCommand() {
        const tsxArgsArray = [];
        // if (!clear && watch) tsxArgsArray.push("--watch-preserve-output");
        // else if (watch) tsxArgsArray.push("--watch");
        tsxArgsArray.push(file);
        if (tsxArgs.trim()) {
            const additionalArgs = tsxArgs
                .trim()
                .split(/\s+/)
                .filter((arg) => arg.length > 0);
            const uniqueArgs = Array.from(new Set([...tsxArgsArray, ...additionalArgs]));
            tsxArgsArray.length = 0;
            tsxArgsArray.push(...uniqueArgs);
        }
        const tsxCommand = `npx tsx ${tsxArgsArray.join(" ")}`;
        if (exports.tsxKiller)
            ((0, exports.tsxKiller)?.()).then(() => (exports.tsxKiller = (0, runner_1.run)(tsxCommand)));
        else
            exports.tsxKiller = (0, runner_1.run)(tsxCommand);
    }
    runTsxCommand();
    if (!typeCheck)
        return;
    const tscArgsArray = [];
    const nodeArgs = maxNodeMem ? [`--max_old_space_size=${maxNodeMem}`] : [];
    tscArgsArray.push((0, compiler_provider_1.getCompilerPath)(compiler));
    tscArgsArray.push("--noEmit");
    if (watch)
        tscArgsArray.push("--watch");
    if (tscArgs.trim()) {
        const additionalArgs = tscArgs
            .trim()
            .split(/\s+/)
            .filter((arg) => arg.length > 0);
        const uniqueArgs = Array.from(new Set([...tscArgsArray, ...additionalArgs]));
        tscArgsArray.length = 0;
        tscArgsArray.push(...uniqueArgs);
    }
    const tscProcess = (0, cross_spawn_1.default)("node", [...nodeArgs, ...tscArgsArray]);
    if (!tscProcess.stdout)
        throw new Error("Unable to read Typescript stdout");
    if (!tscProcess.stderr)
        throw new Error("Unable to read Typescript stderr");
    tscProcess.on("exit", (_, signal) => {
        if (signal !== null)
            process.kill(process.pid, signal);
    });
    tscProcess.stderr.pipe(process.stderr);
    let compilationId = 0;
    let compilationErrorSinceStart = false;
    let hasTsErrors = false;
    function restartTsx() {
        compilationId++;
        (0, killer_1.killProcesses)(compilationId).then((previousCompilationId) => {
            if (previousCompilationId !== compilationId)
                return;
            if (compilationErrorSinceStart)
                Signal.emitFail();
            else {
                Signal.emitSuccess();
                runTsxCommand();
            }
        });
    }
    if (watch)
        (0, file_watcher_1.setupFileWatcher)(restartTsx);
    const rl = (0, readline_1.createInterface)({ input: tscProcess.stdout });
    rl.on("line", function (line) {
        (0, stdout_manipulator_1.print)(line, {
            clear,
        });
        const state = (0, stdout_manipulator_1.detectState)(line);
        const compilationStarted = state.compilationStarted;
        const compilationError = state.compilationError;
        const compilationCompleteWithoutError = state.compilationCompleteWithoutError;
        if (compilationCompleteWithoutError)
            hasTsErrors = false;
        if (compilationError) {
            hasTsErrors = true;
            compilationId++;
            (0, killer_1.killProcesses)(compilationId).then((previousCompilationId) => {
                if (previousCompilationId !== compilationId)
                    return;
                Signal.emitStarted();
            });
        }
        compilationErrorSinceStart =
            (!compilationStarted && compilationErrorSinceStart) || compilationError;
        if (state.fileEmitted !== null)
            Signal.emitFile(state.fileEmitted);
        if (compilationCompleteWithoutError && !hasTsErrors && !firstTime) {
            compilationId++;
            (0, killer_1.killProcesses)(compilationId).then((previousCompilationId) => {
                if (previousCompilationId !== compilationId)
                    return;
                if (compilationErrorSinceStart)
                    Signal.emitFail();
                else {
                    Signal.emitSuccess();
                    runTsxCommand();
                }
            });
        }
        else if (firstTime && compilationCompleteWithoutError && !hasTsErrors) {
            firstTime = false;
            Signal.emitFirstSuccess();
        }
    });
    if (typeof process.on === "function")
        process.on("message", (msg) => {
            if (msg === "run-on-success-command" && exports.tsxKiller)
                (0, exports.tsxKiller)().then(runTsxCommand);
        });
    const sendSignal = (msg) => process.send && process.send(msg);
    const Signal = {
        emitStarted: () => sendSignal("started"),
        emitFirstSuccess: () => sendSignal("first_success"),
        emitSuccess: () => sendSignal("success"),
        emitFail: () => sendSignal("compile_errors"),
        emitFile: (path) => sendSignal(`file_emitted:${path}`),
    };
    (0, node_cleanup_1.default)((_exitCode, signal) => {
        if (signal)
            tscProcess.kill(signal);
        (0, file_watcher_1.stopFileWatcher)();
        (0, killer_1.killProcesses)(0).then(() => process.exit());
        // don't call cleanup handler again
        (0, node_cleanup_1.uninstall)();
        return false;
    });
}
